Fixes #200, add lambda rewrite for clauses (match, FFI, receive)

lepoetemaudit opened the pull request:
Fixes #200; I realised this issue would affect all the expressions that use `#alpaca_clause{}`, i.e. `match`, `beam` and `receive`, so this fix covers all of these.

Thanks @j14159 for your pointers - they saved me quite a lot of time figuring this one out.

j14159 commented:
Glad I could help, thanks for fixing this!  Looks like the new methods in `test_files/use_lambda.alp` don't get exercised by tests yet - might be worth it?

lepoetemaudit commented:
Good point - I'll add.

lepoetemaudit commented:
So actually testing the returned result of the new additions was a good call - it turns out that the lambda being returned by `lambdaMatch` in the use_lambda.alp is always the one in the last clause, no matter whether true or false is passed into it! I'll investigate.

j14159 commented:
Oh interesting!  Nice catch, thanks :)

lepoetemaudit commented:
I was able to recreate the odd lambda behaviour consistently with this test case;

```elm
module lambda
 
export main

-- apply two functions to `value`, return a tuple of each result
let apply2 f1 f2 value = 
  (f1 value, f2 value)

-- Erroneously returns (3, 3), i.e. the first lambda is overwritten by the second
let main () =
  apply2 (fn x -> x + 1) (fn x -> x + 2) 1
```

I fixed this by incrementing the synth lambda counter in what I think is the right place and added a test for this as well. Looks like this PR fixes two bugs for the price of one! Very glad I added the extra tests now.

j14159 commented:
This is great, thanks for digging in!
